[LLVM][CodeGen][AArch64] Remove bogus lowering of sve_while{gt/ge/hi/hs} intrinsics.

paulwalker-arm opened the pull request:
When fed constant operands we try to lower WHILE intrinsics to PTRUE using a fixed length pattern.  This is not valid for the decrementing variants of WHILE because they construct their result predicate vector by traversing from high->low lanes whereas the incrementing variants and PTRUE traverse from low->high lanes.

Whilst we can still utilise PTRUE by reversing its result I figure replacing a single WHILE with multiple SVE instructions is likely counterproductive.

paulwalker-arm committed:
[LLVM][CodeGen][AArch64] Remove bogus lowering of sve_while{gt/ge/hi/hs} intrinsics.

When fed constant operands we try to lower WHILE intrinsics to
PTRUE using a fixed length pattern.  This is not valid for the
decrementing variants of WHILE because they construct their result
predicate vector by traversing from high->low lanes whereas the
incrementing variants and PTRUE traverse from low->high lanes.

Whilst we can still utilise PTRUE by reversing its result I figure
replacing a single WHILE with multiple SVE instructions is likely
counterproductive.

dtemirbulatov reviewed (approved):
LGTM, With unused tests now could be removed.

paulwalker-arm committed:
Remove redundant negative tests.

I've kept just a single test to show we don't perform the ptrue
transformation.

dtemirbulatov reviewed (changes requested):
Sorry, While revisiting this review once again examining the change, I am a bit confused now. If we look at WHILE{GT/GE/HI,HS} instructions, according to the spec it says:
...
for e = elements-1 downto 0
...
Looks like, it should not be a problem for traversing from high->low lanes?

paulwalker-arm commented:
> Sorry, While revisiting this review once again examining the change, I am a bit confused now. If we look at WHILE{GT/GE/HI,HS} instructions, according to the spec it says: ... for e = elements-1 downto 0 ... Looks like, it should not be a problem for traversing from high->low lanes?

The fact the WHILE{GT/GE/HI,HS} instructions traverse from high->low lanes is what makes the replacement with PTRUE incorrect.  For example:
```
whilege(2,0) -> 0 0 0 0 0 1 1 1
```
whereas:
```
ptrue vl3    -> 1 1 1 0 0 0 0 0
```

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5036,20 +5036,17 @@ static inline SDValue getPTrue(SelectionDAG &DAG, SDLoc DL, EVT VT,
 }
 
 static SDValue optimizeWhile(SDValue Op, SelectionDAG &DAG, bool IsSigned,
```
minor nit: is it worth renaming this to `optimizeIncrementingWhile` to distinguish it further?

paulwalker-arm committed:
Rename optimizeWhile as optimizeIncrementingWhile to better reflect its intended use.
